pkg/river/internal/rivertags/rivertags.go

```diff
@@ -0,0 +1,195 @@
+		attrs = append(attrs, "block")
+	}
+	if f&FlagOptional != 0 {
+		attrs = append(attrs, "optional")
+	}
+	if f&FlagLabel != 0 {
+		attrs = append(attrs, "label")
+	}
+
+	return fmt.Sprintf("Flags(%s)", strings.Join(attrs, ","))
+}
+
+// GoString returns the %#v format of Flags.
+func (f Flags) GoString() string { return f.String() }
+
+// Field is a tagged field within a struct.
+type Field struct {
+	Name  string // Name of tagged field
+	Index []int  // Index into field (reflect.Value.FieldByIndex)
+	Flags Flags  // Flags assigned to field
+}
+
+// Get returns the list of tagged fields for some struct type ty. Get panics if
+// ty is not a struct type.
+//
+// Get examines each tagged field in ty for a rvr key. The rvr key is then
+// parsed as containing a name for the field, followed by a required
+// comma-separated list of options. The name may be empty for fields which do
+// not require a name. Get will ignore any field that is not tagged with a rvr
+// key.
+//
+// Get will treat anonymous struct fields as if the inner fields were fields in
+// the outer struct.
+//
+// Examples of struct field tags and their meanings:
+//
+//     // Field is used as a required block named "my_block".
+//     Field struct{} `rvr:"my_block,block"`
+//
+//     // Field is used as an optional block named "my_block".
+//     Field struct{} `rvr:"my_block,block,optional"`
+//
+//     // Field is used as a required attribute named "my_attr".
+//     Field string `rvr:"my_attr,attr"`
+//
+//     // Field is used as an optional attribute named "my_attr".
+//     Field string `rvr:"my_attr,attr,optional"`
+//
+//     // Field is used for storing the label of the block which the struct
+//     // represents.
+//     Field string `rvr:",label"`
+//
+// With the exception of the `rvr:",label"` tag, all tagged fields must have a
+// unique name.
+//
+// The type of tagged fields may be any Go type, with the exception of
+// `rvr:",label"` tags, which must be strings.
+func Get(ty reflect.Type) []Field {
+	if k := ty.Kind(); k != reflect.Struct {
+		panic(fmt.Sprintf("rivertags: Get requires struct kind, got %s", k))
+	}
+
+	var (
+		fields []Field
+
+		usedNames      = make(map[string][]int)
+		usedLabelField = []int(nil)
+	)
+
+	for _, field := range reflect.VisibleFields(ty) {
+		// Skip over anonymous struct fields; they'll be iterated over on future
+		// loops.
+		if field.Anonymous && field.Type.Kind() == reflect.Struct {
+			continue
+		}
+
+		tag, tagged := field.Tag.Lookup("rvr")
+		if !tagged {
+			continue
+		}
+
+		if !field.IsExported() {
+			panic(fmt.Sprintf("river: rvr tag found on unexported field at %s", printPathToField(ty, field.Index)))
+		}
+
+		options := strings.SplitN(tag, ",", 2)
+		if len(options) == 0 {
+			panic(fmt.Sprintf("river: unsupported empty tag at %s", printPathToField(ty, field.Index)))
+		}
+		if len(options) != 2 {
+			panic(fmt.Sprintf("river: field %s tag is missing options", printPathToField(ty, field.Index)))
```

Comment:
So we'll not be allowing a tag like this with no options, right?
```
Field struct{} `rvr:"my_block"`
```

Reply:
Yeah, right now you have to specify if it's an attribute or a block.